debian-stuff/katello

```diff
@@ -134,6 +138,58 @@ class katello_method(pkg_acquire_method):
             protocol = 'http://'
         return protocol + result.group('url')
 
+
+    def __get_rhsm_proxy(self):
+        """
+        Return the proxy server config from /etc/rhsm/rhsm.conf as dict
+        Adapted from https://github.com/Katello/katello-client-bootstrap/blob/master/bootstrap.py
+	"""
+        rhsmconfig = SafeConfigParser()
+        rhsmconfig.read('/etc/rhsm/rhsm.conf')
+
+        proxy_options = [option for option in rhsmconfig.options('server') if option.startswith('proxy')]
+        proxy_config = {}
+        for option in proxy_options:
+            value = rhsmconfig.get('server', option)
+            if value:
+                proxy_config[option] = value
+
+        proxy = ''
+        protocol = 'http'
+        if 'proxy_hostname' in proxy_config:
+            scheme = ''
+            if 'proxy_scheme' in proxy_config:
+                scheme = '{scheme}://'.format(scheme=proxy_config['proxy_scheme'])
+                if proxy_config['proxy_scheme'] == 'https':
+                    protocol = 'https'
+
+            authpass = ''
+            if 'proxy_user' in proxy_config and 'proxy_password' in proxy_config:
+                authpass = '{proxy_user}:{proxy_password}@'.format(
+                    proxy_user=proxy_config['proxy_user'], 
+                    proxy_password=proxy_config['proxy_password']
+                )
+
+            if 'proxy_port' in proxy_config:
+                proxy = '{scheme}{authpass}{proxy_hostname}:{proxy_port}'.format(
+                    scheme,
+                    authpass,
+                    proxy_hostname=proxy_config['proxy_hostname'],
+                    proxy_port=proxy_config['proxy_port']
+                )
+            else:
+                proxy = '{scheme}{authpass}{proxy_hostname}'.format(
+                    scheme,
+                    authpass,
+                    proxy_hostname=proxy_config['proxy_hostname']
+                )
```

Comment:
This way of composing the proxy URL is not safe. In particular, if either the user name or the password have characters such as ':' or '@', the resulting URL is invalid.

Please use `urlparse.urlunparse()` to properly compose an URL with no need to do all the various string compositions with checks. Also, you must properly percent-encode user name, password, and possibly hostname too, so please use `quote()` for that. You can use `six.moves` to make it compatible with both Python 2 and 3.

Reply:
Completely understand why username and password should be encoded. Thanks.

Regarding urlunparse(), this method would create a URL containing: 'scheme, netloc, url, params, query, fragment'. So, it does not allow to set username and password. It would still be necessary to create username:password somehow. Therefore I think its easier to create the URL with .format() as shown in the current state of this PR:

Reply:
'netloc' is actually user+pass+host+port. So yes, unfortunately you need to compose some parts of the URL manually, however it should be still easier than composing all of it.
Also, composing a 'netloc' can be done in gradual steps, rather than duplicating code: for example, your code for the "proxy_port is specified" and "proxy_port is not specified" is almost the same...

Reply:
Please have a look at the latest state of the PR. I changed a lot.

The proxy URL consists of "scheme, username, password, hostname, port" and netloc needs "username, password, hostname, port" - which means, only the "scheme" part would be handled automatically by urlunparse. It would still be necessary to manually construct the netloc from the elements username, password, hostname, port.

Reply:
OK, this new approach seems better. Thanks for the fixes!